slack/management/commands/fetch_slack_activity.py

```diff
@@ -0,0 +1,314 @@
+                    # channel message. Slack always returns the first message
+                    # even if it's older than the oldest we requested.
+                    if thread.last_update_ts is None:
+                        # However, still record that this thread was updated.
+                        # I think this will only will only matter if all
+                        # messages in the thread have been deleted.
+                        thread.last_update_ts = message["ts"]
+                    continue
+
+                # We never need to look at this message again. Oldest messages
+                # come first unlike for channels.
+                thread.last_update_ts = message["ts"]
+
+                if not should_track_message(message):
+                    continue
+
+                if debug:
+                    thread.channel.seenmessage_set.create(
+                        ts=message["ts"],
+                        thread=thread,
+                    )
+                user = get_or_create_user(message["user"])
+                SlackActivityBucket.track_activity(thread.channel, user, message["ts"])
+            thread.save()
+
+
+def locked(fn):
+    @functools.wraps(fn)
+    def inner(*args, **kwargs):
+        cur = connection.cursor()
+        ID = 1028307  # random number to identify this command
+        cur.execute("SELECT pg_try_advisory_lock(%s);", [ID])
+        (got_lock,) = cur.fetchone()
+        if not got_lock:
+            raise CommandError(
+                "Could not obtain lock: "
+                "another instance of this command must be running."
+            )
+        try:
+            return fn(*args, **kwargs)
+        finally:
+            cur.execute("SELECT pg_advisory_unlock(%s);", [ID])
+
+    return inner
+
+
+@click.command()
+@click.argument("channels", nargs=-1)
+@click.option(
+    "--debug",
+    is_flag=True,
+    help=(
+        "Store all messages seen to be able to "
+        "detect bugs (uses lots of database space)."
+    ),
+)
+@locked
+def command(channels, debug):
+    """
+    Download slack activity from channels the bot is a member of.
+
+    CHANNELS is an optional list of channel names (without the #) to limit to.
+    If not provided, all channels the bot is a member of will be fetched.
+
+    This is resumable -- it can be interrupted and restarted without losing
+    progress.
+
+    Do not run multiple instances of this command in parallel.
+    """
+
+    channels = set(channels)
+    selected_channels = []
+    if channels:
+        for channel_data in get_my_channels():
+            if channel_data["name"] in channels:
+                selected_channels.append(channel_data)
+                channels.remove(channel_data["name"])
+        if channels:
+            raise click.BadParameter(
+                "Could not find channels %r (maybe the bot isn't a member?)" % channels
+            )
```

Comment:
Nit pick, but for simple string interpolation like this, we should prefer f-strings.

```suggestion
            raise click.BadParameter(
                f"Could not find channels {channels} (maybe the bot isn't a member?)"
            )
```